### PR TITLE
Absolute luminosity in E906 legacy generator

### PR DESCRIPTION
**Pull request by knakano0524**
This update is to make the E906 legacy generator output the (correct) absolute luminosity to DST.  Actually it is not yet complete.  I want to hear your opinions about implementation.

First I made `SQPrimaryParticleGen` create a `PHGenIntegral` node in DST, which holds the integrated luminosity and other info.  The PYTHIA generator creates it also.  So analyzer can extract the info using a common code to both.  

But the luminosity at present is not 100% correct, which is set by `SQPrimaryVertexGen::traverse()`, because the weight density is used instead of the number density.  But note that this unit difference is cancelled out when analyzer normalizes a weighted event yield by the simulated integrated luminosity.

I added and removed several constant terms of the cross-section formulae in `SQPrimaryParticleGen::generateDrellYan()`.  It's helpful if you could check if the changes are correct.

I also changed the "cross section" value stored in `SQMCEvent`.  It was actually "cross section * luminosity".  It is now
"cross section" and instead the "weight" value is "cross section * luminosity".  So analyzer has to call `get_weight()` to get the event weight.

Just to clarify,
we need the absolute luminosity when estimating an absolute rate, like N of dimuons per spill, month, year, etc.  I think we do need it now or soon, in order to make detailed estimates for the coming beam time.  The PYTHIA generator can already output it, and the legacy generator is helpful for crosscheck and fast estimate.

### CODE REVIEW

**Review by knakano0524 — COMMENTED** [review]
I made this review to write down some remarks.  I would ask you to pay attention to them particularly.

You need not verify the whole code at present, since I will make more changes based on your feedbacks.

**Comment by abinashpun** [comment]
Dear Kenichi,

Thank you for the changes. The codes look fine to me and I also built and ran the branch successfully. I think Kun can comment more on the cross-section calculation part. As far as integral node is concerned, I guess the number of generator-accepted events and processed events is equal for this generator.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C [hunk 1]
```diff
@@ -270,15 +316,19 @@ int SQPrimaryParticleGen::generateDrellYan(PHCompositeNode *topNode,TVector3 vtx
   double xsec_phsp = dim_x1*dim_x2/(dim_x1 + dim_x2)/dim_mass/dim_mass/dim_mass;
   
   //generation limitation
-  double xsec_limit = (massMax - massMin)*(xfMax - xfMin)*(cosThetaMax*cosThetaMax*cosThetaMax/3. 
-							   +cosThetaMax - cosThetaMin*cosThetaMin*cosThetaMin/3. 
-							   - cosThetaMin)*4./3.;
+  //double xsec_limit = (massMax - massMin)*(xfMax - xfMin)*(cosThetaMax*cosThetaMax*cosThetaMax/3. 
+  //      						   +cosThetaMax - cosThetaMin*cosThetaMin*cosThetaMin/3. 
+  //							   - cosThetaMin)*4./3.;
+  double xsec_limit = (pow(cosThetaMax, 3)/3 + cosThetaMax - pow(cosThetaMin, 3)/3 - cosThetaMin) * 4./3.; // or 3/4?
```
[the remark this answers is not in the document]
Comment (abinashpun) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C:322]: I took it from DPSim package of Kun and there is 4/3.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C [hunk 2]
```diff
@@ -200,10 +230,26 @@ int SQPrimaryParticleGen::process_event(PHCompositeNode* topNode)
   TVector3 vtx;
   vtx.SetXYZ(x_vtx,y_vtx,z_vtx);
 
-  if (_DrellYanGen) generateDrellYan(topNode,vtx, pARatio, luminosity);
-  if (_Pythia) generatePythia(topNode,vtx, pARatio);
-  if (_JPsiGen) generateJPsi(topNode,vtx, pARatio, luminosity);
-  if (_PsipGen) generatePsip(topNode,vtx, pARatio, luminosity);
+  int ret; // KN: Do we need to return ABORTEVENT or repeat to generate a good event???
```
[the remark this answers is not in the document]
Comment (abinashpun) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C:233]: I think this is good idea.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C [hunk 3]
```diff
@@ -270,15 +316,19 @@ int SQPrimaryParticleGen::generateDrellYan(PHCompositeNode *topNode,TVector3 vtx
...
   //generation limitation
-  double xsec_limit = (massMax - massMin)*(xfMax - xfMin)*(cosThetaMax*cosThetaMax*cosThetaMax/3. 
-							   +cosThetaMax - cosThetaMin*cosThetaMin*cosThetaMin/3. 
-							   - cosThetaMin)*4./3.;
+  //double xsec_limit = (massMax - massMin)*(xfMax - xfMin)*(cosThetaMax*cosThetaMax*cosThetaMax/3. 
+  //      						   +cosThetaMax - cosThetaMin*cosThetaMin*cosThetaMin/3. 
+  //							   - cosThetaMin)*4./3.;
+  double xsec_limit = (pow(cosThetaMax, 3)/3 + cosThetaMax - pow(cosThetaMin, 3)/3 - cosThetaMin) * 4./3.; // or 3/4?
+
+  const double xsec_const = 8.0 / 9.0 * pi * pow(fine_structure_const, 2) * pow(hbar_Planck, 2) * c_squared;
```
[the remark this answers is not in the document]
Comment (abinashpun) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C:324]: image
This the formula from fermilab paper from p+Cu Collision. Phys. Rev., D43:2815–2836, 1991.
I was wondering if we should have the hbar term.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (knakano0524) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C:233]: Thank you for the checks.  I will modify the behavior of generateDrellYan() etc.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (knakano0524) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C:324]: I just took the code from the E906 generator, and you are correct that "(hbar*c)^2" doesn't make the dimension correct.  I will think of it further.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C (hunk 1 above)
[the remark this answers is not in the document]
Comment (knakano0524) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C:322]: I computed the integral of dN/dOmega = (3/16pi)*(1+cos2theta).  "3/4" is more correct than "4/3", but actually the factor must be "3/8" in the E1039 code, where `cosThetaMin` is -1 or negative values.  In the E906 code, `cosThetaMin` is always 0 practically and the real minimum of cos(theta) is "-cosThetaMax".

Therefore I will change the factor to "3/8".  Having this value, `xsec_limit` becomes 1 when cosThetaMax = 1 and cosThetaMin = -1, reasonably.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (knakano0524) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C:324]: I checked the units of the constants that are defined by Geant4 and CLHEP and used in this code.  They are the natural (Planck) units with MeV and mm.  Thus the use of "(hbar*c)^2" is correct.  But the mass unit in the cross-section formula has to be MeV, not GeV.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C [hunk 4]
```diff
@@ -270,15 +316,19 @@ int SQPrimaryParticleGen::generateDrellYan(PHCompositeNode *topNode,TVector3 vtx
   double xsec_phsp = dim_x1*dim_x2/(dim_x1 + dim_x2)/dim_mass/dim_mass/dim_mass;
   
   //generation limitation
-  double xsec_limit = (massMax - massMin)*(xfMax - xfMin)*(cosThetaMax*cosThetaMax*cosThetaMax/3. 
-							   +cosThetaMax - cosThetaMin*cosThetaMin*cosThetaMin/3. 
-							   - cosThetaMin)*4./3.;
+  //double xsec_limit = (massMax - massMin)*(xfMax - xfMin)*(cosThetaMax*cosThetaMax*cosThetaMax/3.
```
[the remark this answers is not in the document]
Comment (knakano0524) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C]: These terms are kept omitted.  Without them, the newly committed version shows the correct absolute cross-section value.

File: generators/E906LegacyGen/SQPrimaryParticleGen.C [hunk 5]
```diff
@@ -270,15 +316,19 @@ int SQPrimaryParticleGen::generateDrellYan(PHCompositeNode *topNode,TVector3 vtx
...
+  //      						   +cosThetaMax - cosThetaMin*cosThetaMin*cosThetaMin/3. 
+  //							   - cosThetaMin)*4./3.;
+  double xsec_limit = (pow(cosThetaMax, 3)/3 + cosThetaMax - pow(cosThetaMin, 3)/3 - cosThetaMin) * 4./3.; // or 3/4?
+
+  const double xsec_const = 8.0 / 9.0 * pi * pow(fine_structure_const, 2) * pow(hbar_Planck, 2) * c_squared;
   
   //Total cross-section
-  double xsec = xsec_pdf*xsec_kfactor*xsec_phsp*xsec_limit*luminosity;
-  //@cross_section
+  double xsec   = xsec_pdf * xsec_kfactor * xsec_phsp * xsec_limit * xsec_const;
```
[the remark this answers is not in the document]
Comment (knakano0524) [reply generators/E906LegacyGen/SQPrimaryParticleGen.C]: This change is kept.  It has to be announced to all users when this PR is merged.

**Comment by knakano0524** [comment]
I made another commit.  Now the absolute value of the Drell-Yan cross section should be correct.  I modified the structure of `SQPrimaryParticleGen` and added a macro, `CheckParticleGen.C`, so that the computed cross section can be drawn.  Attached is a result.  It agrees with, for example, the NLO calculation shown in page 80 of DocDB 8588 (Shivangi's thesis) within 50%.

xsec_gen_dy

**Comment by knakano0524** [comment]
I would like to merge this PR tomorrow.  Please let me know any further comments if you have.

As a sanity check, I compared the kinematic distributions generated by the default version and this modified version.  All generated events (i.e. no acceptance nor reconstruction) by `e1039-analysis/SimChainDev` were used.  The distribution shapes of both the non-weighted and weighted events stay identical.

[slide.pdf](https://github.com/E1039-Collaboration/e1039-core/files/6572808/slide.pdf)

**Comment by abinashpun** [comment]
> I would like to merge this PR tomorrow. Please let me know any further comments if you have.
> 
> As a sanity check, I compared the kinematic distributions generated by the default version and this modified version. All generated events (i.e. no acceptance nor reconstruction) by `e1039-analysis/SimChainDev` were used. The distribution shapes of both the non-weighted and weighted events stay identical.
> 
> [slide.pdf](https://github.com/E1039-Collaboration/e1039-core/files/6572808/slide.pdf)

Dear Kenichi,

I have no further comments from my side.

**Pull request merged**